test/suite/wthooks.py

```diff
@@ -0,0 +1,216 @@
+# software under copyright law.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+# IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+#
+# [TEST_TAGS]
+# ignored_file
+# [END_TAGS]
+#
+# WiredTigerHookManager
+#   Manage running of hooks
+#
+from __future__ import print_function
+
+import os, sys
+from importlib import import_module
+import wiredtiger
+
+# Three kinds of hooks available:
+HOOK_REPLACE = 1     # replace the call with the hook function
+HOOK_NOTIFY = 2      # call the hook function after the function
+HOOK_ARGS = 3        # transform the arg list before the call
+
+def tty(message):
+    from wttest import WiredTigerTestCase
+    WiredTigerTestCase.tty(message)
+
+# A global function, when called, has no self.  We'll make a "self"
+# which will be the wiredtiger module
+def hooked_function_for_global_function(orig_func, hooks_name, *args):
+    return hooked_function(wiredtiger, orig_func, hooks_name, *args)
+
+def hooked_function(self, orig_func, hooks_name, *args):
+    hook_func_list = getattr(self, hooks_name)
+
+    notifies = []
+    replace_func = None
+
+    # The three kinds of hooks are acted upon at different times.
+    # Before we call the function, we modify the args as indicated
+    # by hooks.  Then we call the function, possibly with a replacement.
+    # Finally, we'll call any notify hooks.
+    #
+    # We only walk through the hook list once, and process the config
+    # hooks while we're doing that, and copy any other hooks needed.
+    for hook_type,hook_func in hook_func_list:
+        if hook_type == HOOK_NOTIFY:
+            notifies.append(hook_func)
+        elif hook_type == HOOK_REPLACE:
+            replace_func = hook_func
+        elif hook_type == HOOK_ARGS:
+            # The arg list may be completely transformed,
+            # and multiple hooks may do this.
+            args = hook_func(self, args)
+
+    if replace_func == None:
+        if self == wiredtiger:
+            ret = orig_func(*args)
+        else:
+            ret = orig_func(self, *args)
+    else:
+        if self == wiredtiger:
+            ret = replace_func(*args)
+        else:
+            ret = replace_func(self, *args)
+
+    for hook_func in notifies:
+        hook_func(ret, self, *args)
+    return ret
+
+class WiredTigerHookManager(object):
+    def __init__(self, hooknames = []):
+        self.hooks = []
+        names_seen = []
+        for name in hooknames:
+            # The hooks are indicated as "somename=arg" or simply "somename".
+            # hook_somename.py will be imported, and initialized with the arg.
```

Comment:
I think this needs to be repeated or more prominently displayed at the top of this file so that someone looking at adding a hook can find it easily in here.

Reply:
I've added a big comment section in `wthooks.py` that also mentions this, among other internal details.